pkg/common-controller/snapshot_controller_base.go

```diff
@@ -407,6 +413,13 @@ func (ctrl *csiSnapshotCommonController) updateContent(content *crdv1.VolumeSnap
 func (ctrl *csiSnapshotCommonController) deleteSnapshot(snapshot *crdv1.VolumeSnapshot) {
 	_ = ctrl.snapshotStore.Delete(snapshot)
 	klog.V(4).Infof("snapshot %q deleted", utils.SnapshotKey(snapshot))
+	driverName, err := ctrl.getSnapshotDriverName(snapshot)
+	if err != nil {
+		klog.Errorf("failed to getSnapshotDriverName while recording metrics for snapshot %q: %s", utils.SnapshotKey(snapshot), err)
+	} else {
+		deleteOperationKey := metrics.NewOperationKey(metrics.DeleteSnapshotOperationName, snapshot.UID)
+		ctrl.metricsManager.RecordMetrics(deleteOperationKey, metrics.NewSnapshotOperationStatus(metrics.SnapshotStatusTypeSuccess), driverName)
```

Comment:
Similar question for the createsnapshot operations

Reply:
Yes, I'm not sure how we can address that short of occasionally checking and cleaning up keys for snapshots that no longer exist. We could run that as a scheduled goroutine. 

How do we handle this in the snapshotStore and contentStore caches @yuxiangqian? Are those cache records cleared on force deletes?

Reply:
My understanding is even a volumesnapshot has been deleted by user forcefully, the controller would still receive an event from API server. The controller should still have the snapshot locally cached in 'snapshotStore' (controller crash will free any RAM allocated anyways), thus in this case there should be no leaking as line 420 will be executed?

@ggriffiths as I suggested previously to remove the "else" block for metric reporting though. otherwise if the snapshotclass has been deleted, it will be a leak.

Reply:
Do you think it would be worth adding some metrics into the operation cache so we could potentially detect leaks? I'm worried this way of metrics caching is prone to leaking especially if we introduce/change logic flow in the future.

Reply:
Yes, I think that could be a good idea. I'm thinking that might be a good followup issue, to keep this PR simple.

Reply:
Let me know if that is okay @msau42 and I'll open a separate issue/PR to do that work. If not, I can add it sometime this week in this PR.

Reply:
Followup is fine, thanks!

Reply:
Sounds good, opened a new issue here: https://github.com/kubernetes-csi/external-snapshotter/issues/434